File: Bindings/dotnet/RoyalApps.RoyalVNCKit/VncCursor.cs
using System;
using System.Diagnostics;
using static RoyalApps.RoyalVNCKit.RoyalVNCKit;

namespace RoyalApps.RoyalVNCKit;

public sealed unsafe class VncCursor : IDisposable
{
    bool _isDisposed;
    void* _instance;
    void* _pixelData;

    internal VncCursor(void* instance)
    {
        ArgumentNullException.ThrowIfNull(instance);
        _instance = instance;
    }

    public bool IsEmpty => rvnc_cursor_is_empty_get(_instance).FromNativeBool();
    public ushort Width => rvnc_cursor_size_width_get(_instance);
    public ushort Height => rvnc_cursor_size_height_get(_instance);
    public ushort HotspotX => rvnc_cursor_hotspot_x_get(_instance);
    public ushort HotspotY => rvnc_cursor_hotspot_y_get(_instance);
    public long BitsPerComponent => rvnc_cursor_bits_per_component_get(_instance);
    public long BitsPerPixel => rvnc_cursor_bits_per_pixel_get(_instance);
    public long BytesPerPixel => rvnc_cursor_bytes_per_pixel_get(_instance);
    public long BytesPerRow => rvnc_cursor_bytes_per_row_get(_instance);

    public ReadOnlySpan<byte> PixelData
    {
        get
        {
            var length = checked((int)rvnc_cursor_pixel_data_size_get(_instance));
            if (length is 0)
                return ReadOnlySpan<byte>.Empty;

            if (_pixelData is null)
                _pixelData = rvnc_cursor_pixel_data_get_copy(_instance);

            Debug.Assert(_pixelData is not null);
            return new(_pixelData, length);
        }
    }

    public void Dispose()
    {
        if (_isDisposed)
            return;

        _isDisposed = true;

        if (_pixelData is not null)
        {
            rvnc_cursor_pixel_data_destroy(_pixelData);
            _pixelData = null;
        }

        _instance = null;
    }
}

File: Bindings/kotlin/RoyalVNCAndroidTest/royalvnc/src/main/java/com/royalapps/royalvnc/VncCursor.kt
package com.royalapps.royalvnc

import android.graphics.Bitmap
import androidx.core.graphics.createBitmap
import com.sun.jna.*
import java.nio.ByteBuffer

data class VncCursor(
    internal val ptr: Pointer /* rvnc_cursor_t */
): AutoCloseable {
    private var _isClosed = false
    private var _pixelData: Pointer? = null

    val empty: Boolean
        get() = RoyalVNCKit.rvnc_cursor_is_empty_get(ptr)

    val width: Short
        get() = RoyalVNCKit.rvnc_cursor_size_width_get(ptr)

    val height: Short
        get() = RoyalVNCKit.rvnc_cursor_size_height_get(ptr)

    val hotspotX: Short
        get() = RoyalVNCKit.rvnc_cursor_hotspot_x_get(ptr)

    val hotspotY: Short
        get() = RoyalVNCKit.rvnc_cursor_hotspot_y_get(ptr)

    val bitsPerComponent: Long
        get() = RoyalVNCKit.rvnc_cursor_bits_per_component_get(ptr)

    val bitsPerPixel: Long
        get() = RoyalVNCKit.rvnc_cursor_bits_per_pixel_get(ptr)

    val bytesPerPixel: Long
        get() = RoyalVNCKit.rvnc_cursor_bytes_per_pixel_get(ptr)

    val bytesPerRow: Long
        get() = RoyalVNCKit.rvnc_cursor_bytes_per_row_get(ptr)

    val pixelData: Pointer?
        get() {
            val size = pixelDataSize

            if (size <= 0) {
                return null
            }

            if (_pixelData == null) {
                _pixelData = RoyalVNCKit.rvnc_cursor_pixel_data_get_copy(ptr)
            }

            return _pixelData
        }

    val pixelDataSize: Long
        get() = RoyalVNCKit.rvnc_cursor_pixel_data_size_get(ptr)

    fun getBitmap(): Bitmap {
        val buffer = ByteBuffer.allocateDirect(pixelDataSize.toInt())

        val bitmap = createBitmap(
            width.toInt(),
            height.toInt(),
            Bitmap.Config.ARGB_8888
        )

        RoyalVNCKit.rvnc_cursor_copy_pixel_data_to_rgba32_buffer(
            ptr,
            buffer
        )

        bitmap.copyPixelsFromBuffer(buffer)

        return bitmap
    }

    override fun close() {
        if (_isClosed) {
            return
        }

        _isClosed = true

        _pixelData?.let {
            RoyalVNCKit.rvnc_cursor_pixel_data_destroy(it)
        }

        _pixelData = null
    }
}
